### PR TITLE
loader: fix multiboot1 ELF loading

### PR DESCRIPTION
**Pull request by phip1611**
Fix a bug in the loader where the loader did weird stuff when 
- filesize > memsize
- a non-LOAD segment is loaded into memory

### CODE REVIEW

File: main.c [hunk 1]
```diff
@@ -275,11 +275,23 @@ static void elf_boot(const struct boot_params params)
...
+                current_program_header->vaddr, current_program_header->paddr,
+                current_program_header->filesize, current_program_header->memsize);
+            continue;
+        } else {
+            printf(
+                "Loading ELF segment (type 0x%x, offset 0x%x, vaddr 0x%x, paddr 0x%x, filesize 0x%x, memsize 0x%x)\n",
+                current_program_header->type, current_program_header->offset,
+                current_program_header->vaddr, current_program_header->paddr,
+                current_program_header->filesize, current_program_header->memsize);
+        }
```
Comment (snue) [inline main.c]: nit: this would also be fine without the `else` branch and cause less code churn.

File: main.c [hunk 2]
```diff
@@ -275,11 +275,23 @@ static void elf_boot(const struct boot_params params)
         const uint8_t *current_elf_segment_in_memory =
             elf_addr_in_memory + current_program_header->offset;
 
-        printf(
-            "Loading ELF segment (type 0x%x, offset 0x%x, vaddr 0x%x, paddr 0x%x, filesize 0x%x, memsize 0x%x)\n",
-            current_program_header->type, current_program_header->offset,
-            current_program_header->vaddr, current_program_header->paddr,
-            current_program_header->filesize, current_program_header->memsize);
+        if (current_program_header->type != 1 /* LOAD segment */) {
```
Comment (snue) [inline main.c]: nit: the comment makes it sound as if this should match on LOAD segments when it does the opposite. The logic is correct though. We only expect to treat loadable segments here and can ignore the rest.

File: main.c [hunk 3]
```diff
@@ -275,11 +275,23 @@ static void elf_boot(const struct boot_params params)
...
+        } else {
+            printf(
+                "Loading ELF segment (type 0x%x, offset 0x%x, vaddr 0x%x, paddr 0x%x, filesize 0x%x, memsize 0x%x)\n",
+                current_program_header->type, current_program_header->offset,
+                current_program_header->vaddr, current_program_header->paddr,
+                current_program_header->filesize, current_program_header->memsize);
+        }
+
+        die_on(current_program_header->filesize > current_program_header->memsize,
+               "filesize > memsize. The ELF segment looks broken.\n");
```
Comment (snue) [inline main.c]: I'm fine with adding this safety check. But have you seen this on a legitimate ELF object or was that a handcrafted experiment?

**Review by snue — COMMENTED** [review]
Looks like a decent fix to me. While at it, could you also fix this wrong region size determination I just spotted a few lines below? https://github.com/cyberus-technology/coreboot-linux-loader/blame/main/main.c#L289
That should just be the `memsize` of course.

I left some more nits and remarks, but nothing that would block merging this change.

File: main.c (hunk 3 above)
Comment (phip1611) [reply main.c]:
```
Program Headers:
  Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz   Flg Align
  LOAD           0x001000 0x0000000000800000 0x0000000000800000 0x00afad 0x00afad   E 0x1000
  LOAD           0x00c000 0x000000000080b000 0x000000000080b000 0x000c08 0x000c08 R   0x1000
  LOAD           0x00d000 0x000000000080c000 0x000000000080c000 0x0070d0 0x1090f0 RW  0x1000
  NOTE           0x0140d0 0x00000000009150f0 0x00000000009150f0 0x000014 0x000000 R   0x4
                                                                ^^^^^^^^^^^^^^^^^
```

However, we filter that out by checking the segment type already.

**Comment by kaldonir** [comment]
Looks good to me! I've ran an internal CI pipeline, and this fixes the issues we have been seeing.

**Pull request merged**